[threaded-animations] scroll-driven animations should only be accelerated if their source is composited

graouts opened the pull request:
#### 87befbbe824b580abd8dca0f8efade4da4c02cf4
<pre>
[threaded-animations] scroll-driven animations should only be accelerated if their source is composited
<a href="https://bugs.webkit.org/show_bug.cgi?id=303136">https://bugs.webkit.org/show_bug.cgi?id=303136</a>
<a href="https://rdar.apple.com/165441462">rdar://165441462</a>

Reviewed by Anne van Kesteren.

Scroll-driven animations make the assumption that their timeline&apos;s source is composited,
exemplified in code with `ASSERT(sourceScrollableArea-&gt;scrollingNodeID())` in the
`ScrollTimeline::createAcceleratedRepresentation()` function. This assumption and that
assertion fail in a number of demos where `overflow: hidden` is set and the source is
not composited.

To address this, we add a new `canBeAccelerated()` function for timelines allowing
acceleration to be opt-in. In the case of document timelines, this is always true.
For progress-based timelines, we ensure their source is composited. Then, in the
`KeyframeEffect::canBeAccelerated()` function we add a check that the animation
timeline can indeed be accelerated.

This acceleration ability is controlled by a cached boolean flag for progress-based
timelines that is updated under `Document::runPostRenderingUpdateAnimationTasks()`.
To do so, we introduce a new `AnimationTimeline::runPostRenderingUpdateTasks()` which,
on top of resetting the accelerated representation as was previously done, updates
that boolean flag, and if it changed, notifies all associated animation effects such
that they may update their accelerated representation.

We test this behavior with a new test that checks that we can make an animation become
accelerated by virtue of toggling the `overflow` property on its timeline&apos;s source.

This uncovered a small issue in `webanimations/threaded-animations/scroll-timeline-in-iframe.html`
where if no content was rendered, the iframe&apos;s document scroller wouldn&apos;t be composited
and the test would fail. We work around this for the moment bug filed bug 303137 to address this.

Test: webanimations/threaded-animations/threaded-and-non-threaded-animations-scroll-source-composition-change.html

* LayoutTests/webanimations/threaded-animations/resources/scroll-timeline-in-iframe-content.html:
* LayoutTests/webanimations/threaded-animations/scroll-timeline-in-iframe.html:
* LayoutTests/webanimations/threaded-animations/threaded-and-non-threaded-animations-scroll-source-composition-change-expected.txt: Added.
* LayoutTests/webanimations/threaded-animations/threaded-and-non-threaded-animations-scroll-source-composition-change.html: Added.
* LayoutTests/webanimations/threaded-animations/threaded-animations-utils.js:
* Source/WebCore/animation/AnimationTimeline.cpp:
(WebCore::AnimationTimeline::runPostRenderingUpdateTasks):
* Source/WebCore/animation/AnimationTimeline.h:
(WebCore::AnimationTimeline::canBeAccelerated const):
(WebCore::AnimationTimeline::computeCanBeAccelerated const):
(WebCore::AnimationTimeline::clearAcceleratedRepresentation): Deleted.
* Source/WebCore/animation/AnimationTimelinesController.cpp:
(WebCore::AnimationTimelinesController::runPostRenderingUpdateTasks):
* Source/WebCore/animation/DocumentTimeline.h:
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::canBeAccelerated const):
(WebCore::KeyframeEffect::timelineAccelerationAbilityDidChange):
* Source/WebCore/animation/KeyframeEffect.h:
* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::computeCanBeAccelerated const):
* Source/WebCore/animation/ScrollTimeline.h:

Canonical link: <a href="https://commits.webkit.org/303589@main">https://commits.webkit.org/303589@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/91b8faa512b8186ffe22864eca67e18907aa271e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132912 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5413 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44003 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140447 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/84943 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/134782 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/5809 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5276 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/101631 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/84943 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135858 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/4080 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/119083 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/82430 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/3969 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/1591 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/83680 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/112990 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/37201 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143099 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5082 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/37784 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/110008 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5164 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/4349 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/110188 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/164/builds/3894 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/115347 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/58638 "Failed to checkout and rebase branch from PR 54482") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20595 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5136 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/33701 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/4976 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/68588 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5226 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5094 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/animation/AnimationTimeline.cpp:
```diff
@@ -138,6 +138,20 @@ Ref<AcceleratedTimeline> AnimationTimeline::createAcceleratedRepresentation()
+{
+    m_acceleratedRepresentation = nullptr;
+
+    auto previousCanBeAccelerated = std::exchange(m_canBeAccelerated, computeCanBeAccelerated());
```
Nit: I think `bool` is preferred here per some internal poll I ran at some point.

graouts replied:
Alright, I'll fix this in the commit.

graouts committed:
[threaded-animations] scroll-driven animations should only be accelerated if their source is composited
https://bugs.webkit.org/show_bug.cgi?id=303136
rdar://165441462

Reviewed by Anne van Kesteren.

Scroll-driven animations make the assumption that their timeline's source is composited,
exemplified in code with `ASSERT(sourceScrollableArea->scrollingNodeID())` in the
`ScrollTimeline::createAcceleratedRepresentation()` function. This assumption and that
assertion fail in a number of demos where `overflow: hidden` is set and the source is
not composited.

To address this, we add a new `canBeAccelerated()` function for timelines allowing
acceleration to be opt-in. In the case of document timelines, this is always true.
For progress-based timelines, we ensure their source is composited. Then, in the
`KeyframeEffect::canBeAccelerated()` function we add a check that the animation
timeline can indeed be accelerated.

This acceleration ability is controlled by a cached boolean flag for progress-based
timelines that is updated under `Document::runPostRenderingUpdateAnimationTasks()`.
To do so, we introduce a new `AnimationTimeline::runPostRenderingUpdateTasks()` which,
on top of resetting the accelerated representation as was previously done, updates
that boolean flag, and if it changed, notifies all associated animation effects such
that they may update their accelerated representation.

We test this behavior with a new test that checks that we can make an animation become
accelerated by virtue of toggling the `overflow` property on its timeline's source.

This uncovered a small issue in `webanimations/threaded-animations/scroll-timeline-in-iframe.html`
where if no content was rendered, the iframe's document scroller wouldn't be composited
and the test would fail. We work around this for the moment bug filed bug 303137 to address this.

Test: webanimations/threaded-animations/threaded-and-non-threaded-animations-scroll-source-composition-change.html

* LayoutTests/webanimations/threaded-animations/resources/scroll-timeline-in-iframe-content.html:
* LayoutTests/webanimations/threaded-animations/scroll-timeline-in-iframe.html:
* LayoutTests/webanimations/threaded-animations/threaded-and-non-threaded-animations-scroll-source-composition-change-expected.txt: Added.
* LayoutTests/webanimations/threaded-animations/threaded-and-non-threaded-animations-scroll-source-composition-change.html: Added.
* LayoutTests/webanimations/threaded-animations/threaded-animations-utils.js:
* Source/WebCore/animation/AnimationTimeline.cpp:
(WebCore::AnimationTimeline::runPostRenderingUpdateTasks):
* Source/WebCore/animation/AnimationTimeline.h:
(WebCore::AnimationTimeline::canBeAccelerated const):
(WebCore::AnimationTimeline::computeCanBeAccelerated const):
(WebCore::AnimationTimeline::clearAcceleratedRepresentation): Deleted.
* Source/WebCore/animation/AnimationTimelinesController.cpp:
(WebCore::AnimationTimelinesController::runPostRenderingUpdateTasks):
* Source/WebCore/animation/DocumentTimeline.h:
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::canBeAccelerated const):
(WebCore::KeyframeEffect::timelineAccelerationAbilityDidChange):
* Source/WebCore/animation/KeyframeEffect.h:
* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::computeCanBeAccelerated const):
* Source/WebCore/animation/ScrollTimeline.h:

Canonical link: https://commits.webkit.org/303589@main
